### PR TITLE
Feature/add product categories

### LINKED ISSUE #2187: Edit Products M3 - Product Categories Step 3: add new categories to the product
After [step 1](https://github.com/woocommerce/woocommerce-android/issues/2185) and [step 2](https://github.com/woocommerce/woocommerce-android/issues/2186), we should provide the option to add a new category to the product.

Annotation on 2020-04-03 at 17-46-46

### PR DESCRIPTION
**Pull request by anitaa1990**
Closes #2187 by providing option to add a new product category.

#### Notes
~- This PR is in draft till this [update product categories PR](https://github.com/woocommerce/woocommerce-android/pull/2541) can be merged.~

#### Screenshots
. . . . . 

#### Testing
- Click on a product from the products tab and click on the categories section. Notice the new `Add Category` button displayed.
  - Add new category without any parent category. Ensure that the new category is added to the site successfully and you are redirected back to the categories list screen with this newly added category already selected.
  - Add new category with parent category. Ensure that the new category is added to the site successfully and you are redirected back to the categories list screen with this newly added category already selected.
  - Add a new product category name and click on the back button. Verify that a discard dialog is displayed. Clicking on `Keep Editing` retains the same screen. Clicking on `Discard`, redirects to the previous screen with no changes.
  - When entering a category name that already exists on the site, an error occurs from the API side. Try entering a duplicate category for a site and verify that an error snackbar is displayed.

##### After adding a new product category
- Verify that the `Done` menu button is visible after adding a new category.
- Click on the back button and verify that a discard dialog is displayed. Clicking on `Keep Editing` retains the same screen. Clicking on `Discard`, redirects to the previous screen with no changes. 
- Click on the `Done` menu button after adding a new product category. Notice the newly added category is visible in the Categories section in the Product Detail screen. 
  - Click on `Update` in the Product Detail screen and verify that the product category is updated successfully

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/res/layout/add_product_category_view.xml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent">
+
+    <androidx.constraintlayout.widget.ConstraintLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:background="?attr/colorSurface"
+        android:paddingTop="@dimen/margin_medium_large"
+        android:paddingBottom="@dimen/margin_medium_large">
```
Comment (AmandaRiu) [inline WooCommerce/src/main/res/layout/add_product_category_view.xml:12]: These are using dimens from the login library and should be switched to dimens from `dimens_base.xml`

File: WooCommerce/src/main/res/layout/add_product_category_view.xml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+        <com.google.android.material.button.MaterialButton
+            android:id="@+id/addCategoriesButton"
+            style="@style/Woo.Button.Outlined"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/margin_extra_large"
+            android:layout_marginEnd="@dimen/margin_extra_large"
```
Comment (AmandaRiu) [inline WooCommerce/src/main/res/layout/add_product_category_view.xml:20]: Same as above

File: WooCommerce/src/main/res/layout/skeleton_product_categories_list.xml [hunk 3]
```diff
@@ -6,6 +6,21 @@
...
+        android:background="@color/skeleton_color"
+        android:layout_marginStart="@dimen/margin_extra_large"
+        android:layout_marginTop="@dimen/margin_extra_large"
+        android:layout_marginEnd="@dimen/margin_extra_large"/>
+
+    <View
+        style="@style/Woo.Divider"
+        android:layout_marginStart="@dimen/major_100"
+        android:layout_marginTop="@dimen/margin_extra_large"
+        android:layout_marginBottom="@dimen/margin_extra_large"/>
```
Comment (AmandaRiu) [inline WooCommerce/src/main/res/layout/skeleton_product_categories_list.xml:22]: Use dimens from `dimens_base.xml`

Comment (anitaa1990) [reply WooCommerce/src/main/res/layout/skeleton_product_categories_list.xml]: 🤦‍♀️ I can't believe I missed that! Fixed in c975f78

File: WooCommerce/src/main/res/layout/add_product_category_view.xml (hunk 1 above)
Comment (anitaa1990) [reply WooCommerce/src/main/res/layout/add_product_category_view.xml]: Fixed in c975f78

File: WooCommerce/src/main/res/layout/add_product_category_view.xml (hunk 2 above)
Comment (anitaa1990) [reply WooCommerce/src/main/res/layout/add_product_category_view.xml]: Fixed in 9eb5a76

**Comment by anitaa1990** [comment]
Thanks for the review @AmandaRiu!

> The start margin/padding doesn’t look to be 16 dp:

Good catch! Fixed in 9eb5a76. 


> When adding a new category, lowercase labels come after uppercase. Is this intentional?

Yup, we sort by category name asc when we fetch categories from the API. This is how it is sorted in Core as well so we are just following the same logic.

Screenshot 2020-06-19 at 10 56 13 AM

Ready for another round 👍

**Comment by anitaa1990** [comment]
Thanks for thew review @AmandaRiu!

That's weird because it's working correctly for me 🤔 Would you let me know which device you are testing on so that I can debug what's going on?



In any case, I opened a separate issue [here](https://github.com/woocommerce/woocommerce-android/issues/2572) to tackle this since this PR is pretty big and I would like to get it merged into `develop` before any merge conflicts. Since this feature will only be live to debug users till the M3 release, it should be fine. WDYT?

**Comment by anitaa1990** [comment]
Hey @AmandaRiu, I fixed the margin issue in c5b911c and tested in Pixel 2, 3 and 4 emulators.

. . 

Ready for another round 👍

**Pull request merged**